Chap_API_Storage.tex

```diff
@@ -0,0 +1,225 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Chapter: Storage support
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\chapter{Storage Support Definitions}
+\label{chap:api_storage}
+
+Distributed and parallel computing systems are increasingly embracing storage hierarchies to meet the diverse data management needs of applications and other systems software in a cost-effective manner.
+These hierarchies provide access to a number of distinct storage layers, with each potentially composed of different storage hardware (e.g., HDD, SSD, tape, PMEM), deployed at different locations (e.g., on-node, on-switch, on-site, WAN), and designed using different storage paradigms (e.g., file-based, object-based).
+Each of these systems offers unique performance and usage characteristics that storage system users should carefully consider to ensure the most efficient use of storage resources.
+
+PMIx enables users to better understand storage hierarchies by defining attributes that formalize storage system characteristics, state, and other parameters.
+These attributes can be queried by applications, I/O libraries and middleware, and workflow systems to discover available storage resources and to inform on which resources are most suitable for different I/O workload requirements.
+
+%%%%%%%%%%%
+\section{Storage support constants}
+
+\declarestruct{pmix_storage_type_t}
+
+The \refstruct{pmix_storage_type_t} structure is an enumerated type for specifying the type of storage system being used.
+
+\begin{constantdesc}
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_LUSTRE}
+Lustre storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_GPFS}
+GPFS (Spectrum Scale) storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_DATAWARP}
+Datawarp storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_NFS}
+NFS storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_CEPH}
+Ceph storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_EXT4}
+Ext4 storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_XFS}
+XFS storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_TMPFS}
+tmpfs storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_UNKNOWN_FS}
+Unknown file-based storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_DAOS}
+DAOS storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_S3}
+S3 storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_GCP}
+GCP storage system.
+%
+\declareconstitemNEW{PMIX_STORAGE_TYPE_UNKNOWN_OS}
+Unknown object-based storage system.
+%
+\end{constantdesc}
+
+
+\declarestruct{pmix_storage_medium_t}
+
+The \refstruct{pmix_storage_medium_t} structure is a \code{uint16_t} type that defines a set of bit-mask flags for specifying different types of storage mediums. These can be bitwise OR'd together to accommodate storage systems that mix storage medium types.
```

Comment:
Any reason to limit this to uint16_t as opposed to starting out with uint32_t or uint64_t?  With 7 mediums defined in this PR, it's already almost 50% allocated.

Reply:
That's a good point, though I think we should be fine for some time. I would say the list of mediums we have defined already is exhaustive for storage devices on current systems, but always hard to speculate how much it could change going forward.

32-bits probably would have been a better decision for future-proofing, but hopefully it's not too big of a deal to up the size when new medium types are added to the standard?

Reply:
> hopefully it's not too big of a deal to up the size when new medium types are added to the standard?

My understanding of C's unions is pretty shaky.  What happens if the underlying value is `uint32_t`/`uint64_t` and you access it as a `uint16_t`?  Does it just truncate off the most significant digits?  If so, I think that would allow upgrading the number of bits while maintaining backwards-compatibility.

Reply:
Changing sizes can be a problem. I would recommend just going with uint64_t from the start and avoid the issue. Given that users are accessing the values with named constants, it really doesn't make any difference, so best to just leave room.

Reply:
Sounds good, let's just change it to avoid any potential issues in the future.